ledger/src/check_next_block.rs

```diff
@@ -129,6 +134,46 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         Ok(())
     }
 
+    /// Check that leaves in the subdag point to batches in other blocks that are valid.
+    ///
+    /// This does not verify that the batches are signed correctly or that the edges are valid
+    /// (only point to the previous round), as those checks already happend when the node received the batch.
+    fn check_block_subdag_leaves(&self, block: &Block<N>) -> Result<()> {
+        // Check if the block has a subdag.
+        let Authority::Quorum(subdag) = block.authority() else {
+            return Ok(());
+        };
+
+        // Store the IDs of all certificates in this subDAG.
+        // This allows determining which edges point to other subDAGs/blocks.
+        let subdag_certs: HashSet<_> = subdag.certificate_ids().collect();
+        let leaf_edges: HashSet<_> = subdag
+            .certificates()
+            .flat_map(|cert| {
+                cert.previous_certificate_ids().iter().map(|prev_id| (cert.id(), cert.round() - 1, prev_id))
```

Comment:
It seems possible to store many instances of `prev_id` if multiple certificates point to it. 

This means we are doing multiple `get_block_for_certificate(prev_id)` DB lookups in the subsequent iterator.

Reply:
Good catch. I added that for debugging and forgot to remove it. Should be resolved now.